### PR TITLE
Update keras, tf and new model usage, numpy 2.0 updates

### PR DESCRIPTION
**Pull request by JGSweets**
this pr:
* updates to allow keras usage > 3.4
* allows usage of most recent TFs / metal instead of macos
* Updates numpy for v2+

NOTES:
* built off of https://github.com/capitalone/DataProfiler/pull/1202 and will need to rebase once that merges

### CODE REVIEW

File: dataprofiler/plugins/decorators.py [hunk 1]
```diff
@@ -21,7 +22,6 @@ def __inner_factory_function(fn):
         :param fn: Plugin function
         :return: function
         """
-        global plugins_dict
```
Comment (JGSweets) [inline dataprofiler/plugins/decorators.py:24]: global is only needed when a function rebinds a module variable, like plugins_dict = {...}.

**Comment by JGSweets** [comment]
@shania-m this is the final PR that will update TF and keras to be current!

**Comment by JGSweets** [comment]
@shania-m I've updated numpy to allow for 2.0+ as well here!

**Comment by shania-m** [comment]
@JGSweets please let me know when it’s ready for review

**Comment by JGSweets** [comment]
@shania-m fixed all the issues related. Updating mypy was more complicated than expected when trying to handle the numpy update!

**Comment by JGSweets** [comment]
@shania-m it is ready for review!

File: dataprofiler/labelers/character_level_cnn_model.py [hunk 2]
```diff
@@ -475,6 +476,28 @@ def _argmax_threshold_layer(
         # matrix.
         return ThreshArgMaxLayer(threshold, num_labels, default_ind)
 
+    @staticmethod
+    def _compile_loss(model: tf.keras.Model, num_labels: int) -> None:
+        """Compiles the loss for the given model and number of labels."""
+        # Compile the model
+        softmax_output_layer_name = model.output_names[0]
+        # losses = {softmax_output_layer_name: "categorical_crossentropy"}
```
Comment (shania-m) [inline dataprofiler/labelers/character_level_cnn_model.py]: please remove

**Review by shania-m — CHANGES_REQUESTED** [review]

File: dataprofiler/labelers/character_level_cnn_model.py [hunk 3]
```diff
@@ -475,6 +476,28 @@ def _argmax_threshold_layer(
...
+        losses = ["categorical_crossentropy", None, None]
```
Comment (shania-m) [inline dataprofiler/labelers/character_level_cnn_model.py]: Quick question — the loss assignment changed from dict-based (by output name) to list-based (by position):                                    
                                                                                                                                                
  # CharacterLevelCnnModel (3 outputs)                                                                                                          
  losses = ["categorical_crossentropy", None, None]                                                                                             
                                                                                                                                                
  # CharLoadTFModel (2 outputs)                                                                                                                 
  losses = ["categorical_crossentropy", None]                                                                                                   
                                                                                                                                                
  Can you confirm the output ordering is stable and these align correctly with the model outputs? Just want to make sure the positional         
  assignment matches up since the dict approach was order-independent.

**Comment by shania-m** [comment]
Few small things, thanks for the contribution!

File: dataprofiler/labelers/character_level_cnn_model.py (hunk 3 above)
Comment (JGSweets) [reply dataprofiler/labelers/character_level_cnn_model.py]: Good call out! I believe the previous layers were list based which is why in keras 3 it required the list losses. This code matched that, however, like you I prefer the order-independent approach and am looking into the requirements of that and ensuring the backwards compatibility of loading a model that was list based initially.

File: dataprofiler/labelers/char_load_tf_model.py [hunk 4]
```diff
@@ -215,15 +247,34 @@ def load_from_disk(cls, dirpath: str) -> CharLoadTFModel:
...
         loaded_model._model_num_labels = loaded_model.num_labels
         loaded_model._model_default_ind = loaded_model.label_mapping[
             loaded_model._parameters["default_label"]
         ]
+        loaded_model._compile_model(loaded_model.num_labels)
         return loaded_model
 
+    def _compile_model(self, num_labels: int) -> None:
+        """Compile the model with dict-based losses and metrics."""
+        losses = {
```
Comment (JGSweets) [inline dataprofiler/labelers/char_load_tf_model.py:263]: ensure we utilize dict based solution

File: dataprofiler/labelers/char_load_tf_model.py [hunk 5]
```diff
@@ -61,6 +64,35 @@ def __init__(
 
         BaseModel.__init__(self, label_mapping, parameters)
 
+    @classmethod
+    def _create_model_outputs(
+        cls, softmax_output: tf.Tensor, argmax_output: tf.Tensor | None = None
+    ) -> dict[str, tf.Tensor]:
+        """Return normalized dict outputs for training and inference."""
+        if argmax_output is None:
```
Comment (JGSweets) [inline dataprofiler/labelers/char_load_tf_model.py:73]: ensure normalized dict based model outputs

File: dataprofiler/labelers/char_load_tf_model.py [hunk 6]
```diff
@@ -61,6 +64,35 @@ def __init__(
...
+            argmax_output = ArgMaxLayer(name=cls._ARGMAX_OUTPUT)(softmax_output)
+        return {
+            cls._SOFTMAX_OUTPUT: softmax_output,
+            cls._ARGMAX_OUTPUT: argmax_output,
+        }
+
+    @classmethod
+    def _normalize_model_outputs(cls, model: tf.keras.Model) -> tf.keras.Model:
+        """Convert list-style outputs to the normalized dict structure."""
+        return labeler_utils.normalize_tf_model_outputs(
```
Comment (JGSweets) [inline dataprofiler/labelers/char_load_tf_model.py:83]: conversion of previous style for consistency requirement by keras 3

File: dataprofiler/labelers/character_level_cnn_model.py [hunk 7]
```diff
@@ -206,6 +220,9 @@ class CharacterLevelCnnModel(BaseTrainableModel, metaclass=AutoSubRegistrationMe
 
     # boolean if the label mapping requires the mapping for index 0 reserved
     requires_zero_mapping: bool = True
+    _SOFTMAX_OUTPUT = "softmax_output"
```
Comment (JGSweets) [inline dataprofiler/labelers/character_level_cnn_model.py:224]: normalize layer names

File: dataprofiler/labelers/char_load_tf_model.py [hunk 8]
```diff
@@ -29,6 +30,8 @@ class CharLoadTFModel(BaseTrainableModel, metaclass=AutoSubRegistrationMeta):
 
     # boolean if the label mapping requires the mapping for index 0 reserved
     requires_zero_mapping = False
+    _SOFTMAX_OUTPUT = "softmax_output"
```
Comment (JGSweets) [inline dataprofiler/labelers/char_load_tf_model.py:34]: normalize layer names

File: dataprofiler/labelers/char_load_tf_model.py [hunk 9]
```diff
@@ -237,45 +288,28 @@ def _construct_model(self) -> None:
...
-        softmax_layer = self._model.get_layer(softmax_output_layer_name)
 
-        new_softmax_layer = softmax_layer.output
+        new_softmax_layer = softmax_output
         if softmax_layer.weights[0].shape[-1] != num_labels:
             new_softmax_layer = tf.keras.layers.Dense(
-                num_labels, activation="softmax", name="softmax_output"
+                num_labels,
+                activation="softmax",
+                name=self._new_softmax_head_name(),
```
Comment (JGSweets) [inline dataprofiler/labelers/char_load_tf_model.py:308]: allows iteration on layer name due to keras reqs

File: dataprofiler/labelers/char_load_tf_model.py [hunk 10]
```diff
@@ -383,18 +399,37 @@ def fit(
...
+                return_dict=True,
+            )
+            acc_value = next(
+                (value for key, value in model_results.items() if key.endswith("acc")),
+                np.nan,
+            )
+            f1_value = next(
+                (value for key, value in model_results.items() if "f1" in key.lower()),
+                np.nan,
+            )
```
Comment (JGSweets) [inline dataprofiler/labelers/char_load_tf_model.py:420]: due to dict based output

File: dataprofiler/labelers/character_level_cnn_model.py [hunk 11]
```diff
@@ -242,6 +259,54 @@ def __init__(self, label_mapping: dict[str, int], parameters: dict = None) -> No
 
         BaseModel.__init__(self, label_mapping, parameters)
 
+    @classmethod
+    def _create_model_outputs(
```
Comment (JGSweets) [inline dataprofiler/labelers/character_level_cnn_model.py:264]: similar to char_load_tf_model.py but with the threshargmax

File: dataprofiler/labelers/character_level_cnn_model.py [hunk 12]
```diff
@@ -690,18 +750,37 @@ def fit(
...
+                return_dict=True,
+            )
+            acc_value = next(
+                (value for key, value in model_results.items() if key.endswith("acc")),
+                np.nan,
+            )
+            f1_value = next(
+                (value for key, value in model_results.items() if "f1" in key.lower()),
+                np.nan,
+            )
```
Comment (JGSweets) [inline dataprofiler/labelers/character_level_cnn_model.py:772]: due to dict based change

File: dataprofiler/labelers/labeler_utils.py [hunk 13]
```diff
@@ -231,6 +231,64 @@ def get_tf_layer_index_from_name(model: tf.keras.Model, layer_name: str) -> int
     return None
 
 
+def normalize_tf_model_outputs(
```
Comment (JGSweets) [inline dataprofiler/labelers/labeler_utils.py:235]: this allows us backwards compatibility with the list based models.

**Comment by JGSweets** [comment]
@shania-m sorry to add so much more, but this should be safer since it has the dict mapping back!

**Comment by shania-m** [comment]
Thanks for the contributions!
Some non blocking recommendations:
  1. Add an upper bound to numpy — numpy>=1.22.0,<3.0.0 instead of numpy>=1.0.0 to prevent future breakage from numpy 3.
  2. Add an upper bound to keras — keras>3.4.0,<4.0.0 to protect against future Keras major versions.
  3. Add a test for loading old-format models — Verify that _normalize_model_outputs correctly handles models saved with the previous list-style
   output format.
  4. Track numpy private API usage — Add a code comment noting that _histograms_impl is a private module, linking to any numpy discussion about a public replacement.
  5. Update CHANGELOG — Document the numpy 2.0 and keras >3.4 support as a notable change.

**Comment by JGSweets** [comment]
1. Updated.
2. Updated.
3. Added unit tests to address this.
4. Added a comment.
5. Added a  CHANGELOG

Thanks!

**Comment by shania-m** [comment]
> Warning
> 
> **Review the following alerts detected in dependencies.**
> 
> According to your organization's Security Policy, it is recommended to resolve "Warn" alerts. Learn more about [Socket for GitHub](https://socket.dev?utm_medium=gh).
> Action	Severity	Alert  (click "▶" to expand/collapse)
> Warn	[High](https://socket.dev/dashboard/org/CapitalOne/diff-scan/fb93fa10-16b6-463d-98b5-4ecd9aace5bf/alert/QlzpKL-e6SclipJA2kKiS3Hc2nht0YycgksTG8-rSmHk)	
> **License policy violation**: pypi `numpy` under FSFAP
> **License:** FSFAP - The applicable license policy does not permit this license (5) (numpy-2.4.6/vendored-meson/meson/test cases/frameworks/6 gettext/data3/metainfo.its)
> 
> **From:** [requirements.txt](https://github.com/capitalone/DataProfiler/pull/1206/files#diff-4d7c51b1efe9043e44439a949dfd92e5827321b34082903477fd04876edb7552) → `pypi/numpy@2.4.6`
> 
> ℹ Read more on: [This package](https://socket.dev/pypi/package/numpy/overview/2.4.6/tar-gz) | [This alert](https://socket.dev/dashboard/org/CapitalOne/diff-scan/fb93fa10-16b6-463d-98b5-4ecd9aace5bf/alert/QlzpKL-e6SclipJA2kKiS3Hc2nht0YycgksTG8-rSmHk) | [What is a license policy violation?](https://socket.dev/alerts/licenseSpdxDisj)
> 
> > _Next steps:_ Take a moment to review the security alert above. Review
> >   the linked package source code to understand the potential risk. Ensure the
> >   package is not malicious before proceeding. If you're unsure how to proceed,
> >   reach out to your security team or ask the Socket team for help at
> >   `support@socket.dev`.
> > 
> > _Suggestion:_ Find a package that does not violate your license policy or adjust your policy to allow this package's license.
> > 
> >     _Mark the package as acceptable risk_. To ignore this alert only
> >     in this pull request, reply with the comment
> >     `@SocketSecurity ignore pypi/numpy@2.4.6`. You can
> >     also ignore all packages with `@SocketSecurity ignore-all`.
> >     To ignore an alert for all future pull requests, use Socket's Dashboard to
> >     change the [triage state of this alert](https://socket.dev/dashboard/org/CapitalOne/diff-scan/fb93fa10-16b6-463d-98b5-4ecd9aace5bf/alert/QlzpKL-e6SclipJA2kKiS3Hc2nht0YycgksTG8-rSmHk).
> 
> [View full report](https://socket.dev/dashboard/org/CapitalOne/diff-scan/fb93fa10-16b6-463d-98b5-4ecd9aace5bf?tab=alerts&action=error%2Cwarn)

@JGSweets i need to review these before approving

**Comment by JGSweets** [comment]
@shania-m of course! ty for working with me on this!

**Comment by JGSweets** [comment]
After this goes in, what would be the steps needed to make a release? I assume I cannot be part of that, but if so I'm happy to help achieve that as well!

**Comment by JGSweets** [comment]
I realized it might also be good to add py3.12 / py3.13 to the test list, especially since py3.10 reaches EOL this fall.

Will do that in a subsequent PR.

**Comment by JGSweets** [comment]
@shania-m any updates on the license review?

**Comment by shania-m** [comment]
> @shania-m any updates on the license review?

Hi! I’m still working on this

**Review by shania-m — APPROVED** [review]

**Comment by JGSweets** [comment]
Awesome @shania-m, Looking forward to this getting through! Like I mentioned, I'd probably follow-up with py3.12, py3.13 test additions to make sure it is forward functional and then figure out with you how to do a release!

Thanks again for all this help!

**Comment by shania-m** [comment]
> Awesome @shania-m, Looking forward to this getting through! Like I mentioned, I'd probably follow-up with py3.12, py3.13 test additions to make sure it is forward functional and then figure out with you how to do a release!
> 
> Thanks again for all this help!

@JGSweets sounds good, just waiting on a review from @ryanSoley

**Comment by JGSweets** [comment]
@shania-m @ryanSoley any updates?

**Comment by JGSweets** [comment]
@shania-m @ryanSoley any updates here? Would love to get this moving!

**Comment by JGSweets** [comment]
@ryanSoley any chance this will be reviewed soon? Thanks!

**Review by ryanSoley — APPROVED** [review]

**Comment by JGSweets** [comment]
@shania-m are we good to go?! Excited to get this in!

**Comment by JGSweets** [comment]
🎆 ty!

**Comment by shania-m** [comment]
@JGSweets can you please let us know if you are blocked / any timelines needed so we can work on allocating capacity for a release. Thank you!

**Comment by vlee7296** [comment]
Can a new release please be created to include this update? Not being able to use keras > 3.4 is tough from a library dependency management perspective.

**Comment by shania-m** [comment]
@zazulam @ryanSoley

**Comment by vlee7296** [comment]
Any update or ETA of when a new release will be available? Thanks!

**Comment by shania-m** [comment]
@JGSweets @vlee7296 Please let me know if there are any critical timelines and I will get to prioritizing this. Thank you for your patience!

**Comment by vlee7296** [comment]
Thanks @shania-m ! A new release by the end of the month would be good.

**Comment by vlee7296** [comment]
Hi @shania-m , Hope you are doing well. I just wanted to check in to see if you were able to get the resources / capacity needed for publishing a release. Thanks!
